File: implementations/python/computational_geometry_basics.py
"""Computational Geometry Fundamentals: 2D Points, Vectors, Exact Orientation,
Segment Intersection, Polygon Area, and Point-in-Polygon Tests.

Implements Arthur's Two-Layer API:
  - Layer 1: Low-level mathematical predicates using exact Python integer arithmetic.
  - Layer 2: Safe, high-level GeometryEngine with bounding box filtering, polygon checks,
             and distance measurements.
"""

from __future__ import annotations
import math
import unittest
from enum import Enum
from typing import NamedTuple, List, Optional, Tuple


class Orientation(Enum):
    COLLINEAR = 0
    COUNTER_CLOCKWISE = 1  # Left turn
    CLOCKWISE = -1         # Right turn


class PointLocation(Enum):
    OUTSIDE = 0
    ON_BOUNDARY = 1
    INSIDE = 2


class Point2D(NamedTuple):
    x: int | float
    y: int | float

    def __add__(self, other: Point2D) -> Point2D:
        return Point2D(self.x + other.x, self.y + other.y)

    def __sub__(self, other: Point2D) -> Point2D:
        return Point2D(self.x - other.x, self.y - other.y)

    def __mul__(self, scalar: int | float) -> Point2D:
        return Point2D(self.x * scalar, self.y * scalar)

    def dot(self, other: Point2D) -> int | float:
        return self.x * other.x + self.y * other.y

    def cross(self, other: Point2D) -> int | float:
        return self.x * other.y - self.y * other.x


# ============================================================================
# Layer 1: Low-Level Primitives & Exact Predicates
# ============================================================================

def cross_product_exact(a: Point2D, b: Point2D, c: Point2D) -> int | float:
    """Computes cross product of vectors (b - a) and (c - a)."""
    return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x)


def orientation(a: Point2D, b: Point2D, c: Point2D) -> Orientation:
    """Returns orientation of triplet (a, b, c): CCW, CW, or COLLINEAR."""
    cp = cross_product_exact(a, b, c)
    if cp > 0:
        return Orientation.COUNTER_CLOCKWISE
    if cp < 0:
        return Orientation.CLOCKWISE
    return Orientation.COLLINEAR


def on_segment_collinear(p: Point2D, q: Point2D, r: Point2D) -> bool:
    """Checks if point q lies on segment pr, assuming p, q, r are collinear."""
    return (min(p.x, r.x) <= q.x <= max(p.x, r.x) and
            min(p.y, r.y) <= q.y <= max(p.y, r.y))


def segments_intersect(p1: Point2D, q1: Point2D, p2: Point2D, q2: Point2D) -> bool:
    """Determines whether line segments p1q1 and p2q2 intersect."""
    o1 = orientation(p1, q1, p2)
    o2 = orientation(p1, q1, q2)
    o3 = orientation(p2, q2, p1)
    o4 = orientation(p2, q2, q1)

    # General crossing case
    if o1 != o2 and o3 != o4:
        return True

    # Collinear overlapping cases
    if o1 == Orientation.COLLINEAR and on_segment_collinear(p1, p2, q1):
        return True
    if o2 == Orientation.COLLINEAR and on_segment_collinear(p1, q2, q1):
        return True
    if o3 == Orientation.COLLINEAR and on_segment_collinear(p2, p1, q2):
        return True
    if o4 == Orientation.COLLINEAR and on_segment_collinear(p2, q1, q2):
        return True

    return False


def polygon_area_2x(vertices: List[Point2D]) -> int | float:
    """Computes twice the signed area of a simple polygon via Shoelace formula."""
    n = len(vertices)
    if n < 3:
        return 0
    area2 = 0
    for i in range(n):
        j = (i + 1) % n
        area2 += vertices[i].x * vertices[j].y - vertices[j].x * vertices[i].y
    return area2


def point_in_polygon(pt: Point2D, poly: List[Point2D]) -> PointLocation:
    """Tests if pt lies INSIDE, OUTSIDE, or ON_BOUNDARY of poly via Ray-Casting."""
    n = len(poly)
    if n < 3:
        return PointLocation.OUTSIDE

    # Boundary check
    for i in range(n):
        j = (i + 1) % n
        if orientation(poly[i], poly[j], pt) == Orientation.COLLINEAR:
            if on_segment_collinear(poly[i], pt, poly[j]):
                return PointLocation.ON_BOUNDARY

    # Ray casting along horizontal line (y = pt.y, x >= pt.x)
    inside = False
    for i in range(n):
        j = (i + 1) % n
        p = poly[i]
        q = poly[j]

        # Order by y-coordinate
        if p.y > q.y:
            p, q = q, p

        # Half-open vertical interval test: p.y <= pt.y < q.y
        if p.y <= pt.y < q.y:
            # Check if pt is strictly to the left of directed edge p -> q
            if orientation(p, q, pt) == Orientation.COUNTER_CLOCKWISE:
                inside = not inside

    return PointLocation.INSIDE if inside else PointLocation.OUTSIDE


# ============================================================================
# Layer 2: Safe High-Level Geometry Engine
# ============================================================================

class GeometryEngine:
    @staticmethod
    def do_segments_intersect(p1: Point2D, q1: Point2D, p2: Point2D, q2: Point2D) -> bool:
        """High-level segment intersection with bounding box fast rejection."""
        if (max(p1.x, q1.x) < min(p2.x, q2.x) or
            max(p2.x, q2.x) < min(p1.x, q1.x) or
            max(p1.y, q1.y) < min(p2.y, q2.y) or
            max(p2.y, q2.y) < min(p1.y, q1.y)):
            return False
        return segments_intersect(p1, q1, p2, q2)

    @staticmethod
    def polygon_area(vertices: List[Point2D]) -> float:
        """Returns absolute area of simple polygon."""
        return abs(polygon_area_2x(vertices)) / 2.0

    @staticmethod
    def locate_point(pt: Point2D, polygon: List[Point2D]) -> PointLocation:
        """Returns location of point relative to polygon."""
        return point_in_polygon(pt, polygon)

    @staticmethod
    def point_to_segment_distance(pt: Point2D, a: Point2D, b: Point2D) -> float:
        """Calculates Euclidean distance from point pt to segment ab."""
        ab = b - a
        ap = pt - a
        ab_len_sq = ab.x * ab.x + ab.y * ab.y
        if ab_len_sq < 1e-15:
            return math.hypot(ap.x, ap.y)

        t = max(0.0, min(1.0, (ap.x * ab.x + ap.y * ab.y) / ab_len_sq))
        closest = Point2D(a.x + t * ab.x, a.y + t * ab.y)
        return math.hypot(pt.x - closest.x, pt.y - closest.y)

    @staticmethod
    def line_segment_intersection_point(p1: Point2D, q1: Point2D, p2: Point2D, q2: Point2D) -> Optional[Point2D]:
        """Calculates floating-point intersection coordinate of two segments if intersecting."""
        d1 = q1 - p1
        d2 = q2 - p2
        denom = d1.x * d2.y - d1.y * d2.x
        if abs(denom) < 1e-15:
            return None

        dp = p2 - p1
        t = (dp.x * d2.y - dp.y * d2.x) / denom
        u = (dp.x * d1.y - dp.y * d1.x) / denom

        if 0.0 <= t <= 1.0 and 0.0 <= u <= 1.0:
            return Point2D(p1.x + t * d1.x, p1.y + t * d1.y)
        return None


# ============================================================================
# Unit & Differential Tests
# ============================================================================

class TestComputationalGeometryBasics(unittest.TestCase):
    def test_orientation(self):
        a = Point2D(0, 0)
        b = Point2D(4, 0)
        c = Point2D(2, 2)
        self.assertEqual(orientation(a, b, c), Orientation.COUNTER_CLOCKWISE)
        self.assertEqual(orientation(b, a, c), Orientation.CLOCKWISE)
        self.assertEqual(orientation(a, b, Point2D(8, 0)), Orientation.COLLINEAR)

    def test_massive_integer_precision(self):
        # Stress test: 10^18 coordinate values (Python handles arbitrarily large integers)
        p1 = Point2D(10**18, 10**18)
        p2 = Point2D(2 * 10**18, 2 * 10**18)
        p3 = Point2D(3 * 10**18, 3 * 10**18 + 1)
        self.assertEqual(orientation(p1, p2, p3), Orientation.COUNTER_CLOCKWISE)

    def test_segment_intersection(self):
        eng = GeometryEngine()
        # Proper cross
        self.assertTrue(eng.do_segments_intersect(Point2D(0, 0), Point2D(4, 4), Point2D(0, 4), Point2D(4, 0)))
        # Parallel disjoint
        self.assertFalse(eng.do_segments_intersect(Point2D(0, 0), Point2D(4, 4), Point2D(0, 1), Point2D(4, 5)))
        # Collinear overlapping
        self.assertTrue(eng.do_segments_intersect(Point2D(0, 0), Point2D(4, 4), Point2D(2, 2), Point2D(6, 6)))
        # Collinear disjoint
        self.assertFalse(eng.do_segments_intersect(Point2D(0, 0), Point2D(4, 4), Point2D(5, 5), Point2D(8, 8)))
        # T-junction
        self.assertTrue(eng.do_segments_intersect(Point2D(2, 0), Point2D(2, 2), Point2D(0, 2), Point2D(4, 2)))

    def test_polygon_area(self):
        eng = GeometryEngine()
        square = [Point2D(0, 0), Point2D(4, 0), Point2D(4, 4), Point2D(0, 4)]
        self.assertAlmostEqual(eng.polygon_area(square), 16.0)

        triangle = [Point2D(0, 0), Point2D(5, 0), Point2D(0, 6)]
        self.assertAlmostEqual(eng.polygon_area(triangle), 15.0)

    def test_point_in_polygon(self):
        eng = GeometryEngine()
        # L-shaped polygon:
        # (0,0) -> (4,0) -> (4,2) -> (2,2) -> (2,4) -> (0,4)
        l_poly = [
            Point2D(0, 0), Point2D(4, 0), Point2D(4, 2),
            Point2D(2, 2), Point2D(2, 4), Point2D(0, 4)
        ]

        self.assertEqual(eng.locate_point(Point2D(1, 1), l_poly), PointLocation.INSIDE)
        self.assertEqual(eng.locate_point(Point2D(1, 3), l_poly), PointLocation.INSIDE)
        self.assertEqual(eng.locate_point(Point2D(3, 3), l_poly), PointLocation.OUTSIDE)
        self.assertEqual(eng.locate_point(Point2D(2, 2), l_poly), PointLocation.ON_BOUNDARY)
        self.assertEqual(eng.locate_point(Point2D(0, 2), l_poly), PointLocation.ON_BOUNDARY)

    def test_point_to_segment_distance(self):
        eng = GeometryEngine()
        a = Point2D(0, 0)
        b = Point2D(10, 0)
        p = Point2D(5, 5)
        self.assertAlmostEqual(eng.point_to_segment_distance(p, a, b), 5.0)

        p_outside = Point2D(15, 0)
        self.assertAlmostEqual(eng.point_to_segment_distance(p_outside, a, b), 5.0)

    def test_intersection_point(self):
        eng = GeometryEngine()
        pt = eng.line_segment_intersection_point(Point2D(0, 0), Point2D(4, 4), Point2D(0, 4), Point2D(4, 0))
        self.assertIsNotNone(pt)
        self.assertAlmostEqual(pt.x, 2.0)
        self.assertAlmostEqual(pt.y, 2.0)


if __name__ == "__main__":
    unittest.main()

File: implementations/cpp/computational_geometry_basics.cpp
/**
 * @file computational_geometry_basics.cpp
 * @brief Computational Geometry Fundamentals: 2D Points, Vectors, Exact Orientation,
 *        Segment Intersection, Polygon Area, and Point-in-Polygon Tests.
 *
 * Designed with Arthur's Two-Layer API:
 *   - Layer 1: Low-level exact geometric primitives (Point2D, exact cross/dot products, CCW predicate).
 *   - Layer 2: Safe, validated high-level GeometryEngine (segment intersection, ray-casting point-in-polygon,
 *              bounding boxes, polygon area).
 *
 * Implements exact integer arithmetic via __int128_t to avoid floating-point drift and overflow.
 *
 * Compilation:
 *   g++ -std=c++17 -O3 -Wall -Wextra -Werror computational_geometry_basics.cpp -o geom_basics
 */

#include <iostream>
#include <vector>
#include <cmath>
#include <cstdint>
#include <cassert>
#include <optional>
#include <algorithm>
#include <string>
#include <sstream>

namespace geom {

// ============================================================================
// Layer 1: Low-Level Primitives & Exact Predicates
// ============================================================================

template <typename T>
struct Point2D {
    T x{};
    T y{};

    constexpr Point2D() = default;
    constexpr Point2D(T x_val, T y_val) : x(x_val), y(y_val) {}

    constexpr Point2D operator+(Point2D const& other) const {
        return Point2D(x + other.x, y + other.y);
    }

    constexpr Point2D operator-(Point2D const& other) const {
        return Point2D(x - other.x, y - other.y);
    }

    constexpr Point2D operator*(T scalar) const {
        return Point2D(x * scalar, y * scalar);
    }

    constexpr bool operator==(Point2D const& other) const {
        return x == other.x && y == other.y;
    }

    constexpr bool operator!=(Point2D const& other) const {
        return !(*this == other);
    }

    constexpr bool operator<(Point2D const& other) const {
        if (x != other.x) return x < other.x;
        return y < other.y;
    }
};

using Point64 = Point2D<int64_t>;
using PointF = Point2D<double>;

enum class Orientation {
    COLLINEAR = 0,
    COUNTER_CLOCKWISE = 1, // Left turn
    CLOCKWISE = -1         // Right turn
};

enum class PointLocation {
    OUTSIDE = 0,
    ON_BOUNDARY = 1,
    INSIDE = 2
};

struct Segment64 {
    Point64 p1;
    Point64 q1;
};

// Exact 128-bit cross product of vectors (b - a) and (c - a)
inline __int128_t cross_product_exact(Point64 a, Point64 b, Point64 c) {
    __int128_t x1 = static_cast<__int128_t>(b.x) - static_cast<__int128_t>(a.x);
    __int128_t y1 = static_cast<__int128_t>(b.y) - static_cast<__int128_t>(a.y);
    __int128_t x2 = static_cast<__int128_t>(c.x) - static_cast<__int128_t>(a.x);
    __int128_t y2 = static_cast<__int128_t>(c.y) - static_cast<__int128_t>(a.y);
    return x1 * y2 - y1 * x2;
}

// Exact 128-bit dot product of vectors (b - a) and (c - a)
inline __int128_t dot_product_exact(Point64 a, Point64 b, Point64 c) {
    __int128_t x1 = static_cast<__int128_t>(b.x) - static_cast<__int128_t>(a.x);
    __int128_t y1 = static_cast<__int128_t>(b.y) - static_cast<__int128_t>(a.y);
    __int128_t x2 = static_cast<__int128_t>(c.x) - static_cast<__int128_t>(a.x);
    __int128_t y2 = static_cast<__int128_t>(c.y) - static_cast<__int128_t>(a.y);
    return x1 * x2 + y1 * y2;
}

// Orientation predicate: returns CCW, CW, or COLLINEAR
inline Orientation orientation(Point64 a, Point64 b, Point64 c) {
    __int128_t cp = cross_product_exact(a, b, c);
    if (cp > 0) return Orientation::COUNTER_CLOCKWISE;
    if (cp < 0) return Orientation::CLOCKWISE;
    return Orientation::COLLINEAR;
}

// Checks if point q lies on segment pr (assuming p, q, r are collinear)
inline bool on_segment_collinear(Point64 p, Point64 q, Point64 r) {
    return (q.x <= std::max(p.x, r.x) && q.x >= std::min(p.x, r.x) &&
            q.y <= std::max(p.y, r.y) && q.y >= std::min(p.y, r.y));
}

// Tests if two line segments p1q1 and p2q2 intersect
inline bool segments_intersect(Point64 p1, Point64 q1, Point64 p2, Point64 q2) {
    Orientation o1 = orientation(p1, q1, p2);
    Orientation o2 = orientation(p1, q1, q2);
    Orientation o3 = orientation(p2, q2, p1);
    Orientation o4 = orientation(p2, q2, q1);

    // General case: segments straddle each other
    if (o1 != o2 && o3 != o4) {
        return true;
    }

    // Special collinear cases
    if (o1 == Orientation::COLLINEAR && on_segment_collinear(p1, p2, q1)) return true;
    if (o2 == Orientation::COLLINEAR && on_segment_collinear(p1, q2, q1)) return true;
    if (o3 == Orientation::COLLINEAR && on_segment_collinear(p2, p1, q2)) return true;
    if (o4 == Orientation::COLLINEAR && on_segment_collinear(p2, q1, q2)) return true;

    return false;
}

// Doubled signed area of polygon: Shoelace formula
inline __int128_t polygon_area_2x_exact(std::vector<Point64> const& vertices) {
    size_t n = vertices.size();
    if (n < 3) return 0;
    __int128_t area2 = 0;
    for (size_t i = 0; i < n; ++i) {
        size_t j = (i + 1) % n;
        __int128_t xi = vertices[i].x;
        __int128_t yi = vertices[i].y;
        __int128_t xj = vertices[j].x;
        __int128_t yj = vertices[j].y;
        area2 += (xi * yj - xj * yi);
    }
    return area2;
}

// Point-in-polygon test via Ray-Casting with exact boundary semantics
inline PointLocation point_in_polygon_exact(Point64 pt, std::vector<Point64> const& poly) {
    size_t n = poly.size();
    if (n < 3) return PointLocation::OUTSIDE;

    // Check if point lies on any polygon edge
    for (size_t i = 0; i < n; ++i) {
        size_t j = (i + 1) % n;
        if (orientation(poly[i], poly[j], pt) == Orientation::COLLINEAR &&
            on_segment_collinear(poly[i], pt, poly[j])) {
            return PointLocation::ON_BOUNDARY;
        }
    }

    // Ray casting to +x direction (horizontal ray pt.y)
    bool inside = false;
    for (size_t i = 0; i < n; ++i) {
        size_t j = (i + 1) % n;
        Point64 p = poly[i];
        Point64 q = poly[j];

        // Ensure p is lower or equal in y
        if (p.y > q.y) std::swap(p, q);

        // Ray condition: p.y <= pt.y < q.y (half-open vertical interval)
        if (p.y <= pt.y && pt.y < q.y) {
            // Orientation of (p -> q) relative to pt
            // If pt is strictly to the left of directed edge p -> q, ray crosses edge
            Orientation o = orientation(p, q, pt);
            if (o == Orientation::COUNTER_CLOCKWISE) {
                inside = !inside;
            }
        }
    }

    return inside ? PointLocation::INSIDE : PointLocation::OUTSIDE;
}

// ============================================================================
// Layer 2: High-Level Geometric Engine
// ============================================================================

class GeometryEngine {
public:
    static bool doSegmentsIntersect(Point64 p1, Point64 q1, Point64 p2, Point64 q2) {
        // Fast rejection test: Bounding boxes must overlap
        int64_t min_x1 = std::min(p1.x, q1.x), max_x1 = std::max(p1.x, q1.x);
        int64_t min_y1 = std::min(p1.y, q1.y), max_y1 = std::max(p1.y, q1.y);
        int64_t min_x2 = std::min(p2.x, q2.x), max_x2 = std::max(p2.x, q2.x);
        int64_t min_y2 = std::min(p2.y, q2.y), max_y2 = std::max(p2.y, q2.y);

        if (max_x1 < min_x2 || max_x2 < min_x1 || max_y1 < min_y2 || max_y2 < min_y1) {
            return false;
        }

        return segments_intersect(p1, q1, p2, q2);
    }

    static double computePolygonArea(std::vector<Point64> const& polygon) {
        __int128_t area2 = polygon_area_2x_exact(polygon);
        if (area2 < 0) area2 = -area2;
        return static_cast<double>(area2) / 2.0;
    }

    static PointLocation testPointInPolygon(Point64 pt, std::vector<Point64> const& polygon) {
        return point_in_polygon_exact(pt, polygon);
    }

    static double pointToSegmentDistance(PointF p, PointF a, PointF b) {
        double ab_x = b.x - a.x;
        double ab_y = b.y - a.y;
        double ap_x = p.x - a.x;
        double ap_y = p.y - a.y;

        double ab_len_sq = ab_x * ab_x + ab_y * ab_y;
        if (ab_len_sq < 1e-12) {
            return std::hypot(ap_x, ap_y);
        }

        // Projection factor t = (ap . ab) / |ab|^2 clamped to [0, 1]
        double t = (ap_x * ab_x + ap_y * ab_y) / ab_len_sq;
        t = std::max(0.0, std::min(1.0, t));

        double closest_x = a.x + t * ab_x;
        double closest_y = a.y + t * ab_y;
        return std::hypot(p.x - closest_x, p.y - closest_y);
    }

    static std::optional<PointF> lineSegmentIntersectionPoint(PointF p1, PointF q1, PointF p2, PointF q2) {
        double d1_x = q1.x - p1.x;
        double d1_y = q1.y - p1.y;
        double d2_x = q2.x - p2.x;
        double d2_y = q2.y - p2.y;

        double denom = d1_x * d2_y - d1_y * d2_x;
        if (std::abs(denom) < 1e-12) {
            return std::nullopt; // Parallel or collinear
        }

        double dp_x = p2.x - p1.x;
        double dp_y = p2.y - p1.y;

        double t = (dp_x * d2_y - dp_y * d2_x) / denom;
        double u = (dp_x * d1_y - dp_y * d1_x) / denom;

        if (t >= 0.0 && t <= 1.0 && u >= 0.0 && u <= 1.0) {
            return PointF(p1.x + t * d1_x, p1.y + t * d1_y);
        }
        return std::nullopt;
    }
};

} // namespace geom

// ============================================================================
// Verification & Differential Test Suite
// ============================================================================

void run_orientation_tests() {
    using namespace geom;
    Point64 a(0, 0), b(4, 0), c(2, 2);
    assert(orientation(a, b, c) == Orientation::COUNTER_CLOCKWISE);
    assert(orientation(b, a, c) == Orientation::CLOCKWISE);
    assert(orientation(a, b, Point64(8, 0)) == Orientation::COLLINEAR);

    // Coordinate magnitude stress test (no overflow with 64-bit coordinates)
    Point64 p1(1'000'000'000LL, 1'000'000'000LL);
    Point64 p2(2'000'000'000LL, 2'000'000'000LL);
    Point64 p3(3'000'000'000LL, 3'000'000'000LL + 1);
    assert(orientation(p1, p2, p3) == Orientation::COUNTER_CLOCKWISE);
}

void run_segment_intersection_tests() {
    using namespace geom;
    // Proper crossing
    Point64 p1(0, 0), q1(4, 4);
    Point64 p2(0, 4), q2(4, 0);
    assert(GeometryEngine::doSegmentsIntersect(p1, q1, p2, q2) == true);

    // Parallel disjoint
    Point64 p3(0, 1), q3(4, 5);
    assert(GeometryEngine::doSegmentsIntersect(p1, q1, p3, q3) == false);

    // Collinear overlapping
    Point64 p4(2, 2), q4(6, 6);
    assert(GeometryEngine::doSegmentsIntersect(p1, q1, p4, q4) == true);

    // Collinear disjoint
    Point64 p5(5, 5), q5(8, 8);
    assert(GeometryEngine::doSegmentsIntersect(p1, q1, p5, q5) == false);

    // T-junction (endpoint touching interior of segment)
    Point64 p6(2, 0), q6(2, 2);
    Point64 seg_a(0, 2), seg_b(4, 2);
    assert(GeometryEngine::doSegmentsIntersect(p6, q6, seg_a, seg_b) == true);
}

void run_polygon_area_tests() {
    using namespace geom;
    // Standard 4x4 square -> Area = 16.0
    std::vector<Point64> square = {{0, 0}, {4, 0}, {4, 4}, {0, 4}};
    assert(std::abs(GeometryEngine::computePolygonArea(square) - 16.0) < 1e-9);

    // Triangle: (0,0), (5,0), (0,6) -> Area = 15.0
    std::vector<Point64> triangle = {{0, 0}, {5, 0}, {0, 6}};
    assert(std::abs(GeometryEngine::computePolygonArea(triangle) - 15.0) < 1e-9);
}

void run_point_in_polygon_tests() {
    using namespace geom;
    // Non-convex L-shaped polygon
    // (0,0) -> (4,0) -> (4,2) -> (2,2) -> (2,4) -> (0,4)
    std::vector<Point64> l_poly = {
        {0, 0}, {4, 0}, {4, 2}, {2, 2}, {2, 4}, {0, 4}
    };

    // Strictly inside
    assert(GeometryEngine::testPointInPolygon({1, 1}, l_poly) == PointLocation::INSIDE);
    assert(GeometryEngine::testPointInPolygon({1, 3}, l_poly) == PointLocation::INSIDE);

    // Cutout area (outside the reflex notch)
    assert(GeometryEngine::testPointInPolygon({3, 3}, l_poly) == PointLocation::OUTSIDE);

    // On vertices and boundary edges
    assert(GeometryEngine::testPointInPolygon({2, 2}, l_poly) == PointLocation::ON_BOUNDARY);
    assert(GeometryEngine::testPointInPolygon({0, 2}, l_poly) == PointLocation::ON_BOUNDARY);
    assert(GeometryEngine::testPointInPolygon({3, 0}, l_poly) == PointLocation::ON_BOUNDARY);
}

void run_floating_point_distance_tests() {
    using namespace geom;
    PointF a(0, 0), b(10, 0);
    PointF p(5, 5);
    assert(std::abs(GeometryEngine::pointToSegmentDistance(p, a, b) - 5.0) < 1e-9);

    PointF p_outside(15, 0);
    assert(std::abs(GeometryEngine::pointToSegmentDistance(p_outside, a, b) - 5.0) < 1e-9);

    // Intersection point
    auto pt = GeometryEngine::lineSegmentIntersectionPoint({0, 0}, {4, 4}, {0, 4}, {4, 0});
    assert(pt.has_value());
    assert(std::abs(pt->x - 2.0) < 1e-9 && std::abs(pt->y - 2.0) < 1e-9);
}

int main() {
    std::cout << "[Verification] Running Computational Geometry Basics test suite...\n";
    run_orientation_tests();
    run_segment_intersection_tests();
    run_polygon_area_tests();
    run_point_in_polygon_tests();
    run_floating_point_distance_tests();
    std::cout << "[Verification] All 5 test suites passed successfully!\n";
    return 0;
}
